Docs for SCIM2 Build User Creation Payload (#4914)
wso2/docs-is: en/includes/apis/scim2/build-scim2-user-creation-payload.md

# Build SCIM 2.0 User Creation Payloads

> This guide explains how to build SCIM 2.0 payloads in compliance with the SCIM 2.0 specification.

Reviewer: ```suggestion
This guide explains how to build SCIM 2.0 user creation payloads in compliance with the SCIM 2.0 specification.
```

> ## Step 1 : Identifying SCIM 2.0 Claims for User Attributes

Reviewer: ```suggestion
## Step 1 : Identifying SCIM 2.0 Attributes for User Attributes
```

To build a SCIM 2.0 payload, the first step is identifying the SCIM schema mapping for your user attribute:

Navigate to **User Attributes & Stores** → **Attributes** → **SCIM 2.0** in the {{ product_name }} console. You will see the following schema options:

  - **Core Schema**
  - **User Schema**
  - **Enterprise Schema**

>   - **Custom Schema**  {% if product_name == "WSO2 Identity Server" %} (if you have mapped any custom user attribute to a SCIM claim). {% endif %} 

Reviewer: ```suggestion
  - **Custom Schema**  {% if product_name == "WSO2 Identity Server" %} (if you have mapped any custom user attribute to a SCIM attribute in the custom schema). {% endif %} 
```

> For further details refer [Manage SCIM 2.0 attribute mappings]({{base_path}}/guides/users/attributes/manage-scim2-attribute-mappings)

### Rules for Schema Usage in SCIM 2.0 Payloads

- If your user attribute is mapped to **Core Schema** or **User Schema**, the schema name does not need to be qualified in the SCIM payload.
- If your user attribute is mapped to **Enterprise Schema** or **Custom Schema**, each SCIM attribute under these schemas must be placed under the schema’s namespace.

#### Example
```json
{
 "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
   "employeeNumber": "1234A"
 },
 "urn:scim:wso2:schema": {
   "customAttribute": "xyz"
 }
}

> 
> 

Reviewer: ```suggestion
```

```

> ## Step 2 : Attribute Types in SCIM 2.0 Payloads

Reviewer: ```suggestion
## Step 2 : Identify the Attribute Types and their SCIM 2.0 Payloads
```

Each SCIM attribute falls into one of the following attribute types, which determine the format of the attribute in the payload.

### Singular Attributes

1. **Simple Attributes**

    Example:
    ```json
    {
      "userName": "kim"
    }
    ```

2. **Complex Attributes**

    Example:
    ```json
    {
      "name": {
        "givenName": "Kim",
        "familyName": "Berry"
      }
    }
    ```

### Multi-Valued Attributes
1. **Simple Attributes**

    Extended attribute example:
    ```json
    {

>       "urn:scim:wso2:schema:userDevices": {

Reviewer: ```suggestion
      "urn:scim:wso2:schema:user": {
```

2. **Complex Attributes**

    Example for `emails` attribute:
    ```json
    {
      "emails": [
        {
          "type": "home",
          "value": "kim@gmail.com",
          "primary": true
        },
        {
          "type": "work",
          "value": "kim@wso2.com"
        }
      ]
    }
    ```

!!! note
    The following references provide comprehensive information about SCIM attribute types and their respective definitions. These details can help identify the type of attributes used in SCIM 2.0 payloads:

      - For attributes under **Core Schema**, **User Schema** and Specification-defined Enterprise Schema, refer to [RFC 7643 Section 8.7.1](https://datatracker.ietf.org/doc/html/rfc7643#section-8.7.1).

      {% if product_name == "WSO2 Identity Server" %}
      - For WSO2 IS-defined Enterprise Schema attributes, refer to the `scim2-schema-extension.config` file located in the `<IS_HOME>/repository/conf/` directory.
      {% endif %} 

      - For Custom Schema attributes, check the type meta attribute of the mapped local attribute.

## Step 3: Determining the Type of SCIM 2.0 Attributes

> The patterns described below are used for SCIM attributes to map local user attributes. SCIM claim mappings should follow these patterns for different types of SCIM attributes.

Reviewer: ```suggestion
The patterns described below are used for SCIM attributes to map local user attributes. SCIM attribute mappings should follow these patterns for different types of SCIM attributes.
```

### Multivalued Simple Attributes

- **Format:** `<schema name>:<attribute name>`
- These attributes are treated as multivalued simple attributes when the `type` property is specified.

  > By default, Core Schema, User Schema, and Enterprise Schema do not have multivalued simple attributes.

### Complex Attributes with Sub-Attributes

- **Format:** `<schema name>:<attribute name>.<sub attribute>`
- This format is used for sub-attributes of complex attributes.

### Multivalued Complex Attributes

- **Format:** `<schema name>:<multivalued attribute name>` and `<schema name>:<multivalued attribute name>.<type>`
- By default, multivalued complex attributes support only the `type` and `value` sub-attributes.
- Examples include attributes like `emails` and `addresses`.

## Example Payload
```json
{
 "schemas": [
   "urn:ietf:params:scim:schemas:core:2.0:User",
   "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User",
   "urn:scim:wso2:schema"
 ],
 "userName": "kim",
 "password": "MyPa33w@rd",
 "name": {
   "givenName": "Kim",
   "familyName": "Berry"
 },
 "emails": [
   {
     "type": "home",
     "value": "kim@gmail.com",

>      "primary": true

Reviewer: don't use primary and type together since it is not working in IS

   },
   {
     "type": "work",
     "value": "kim@wso2.com"
   }
 ],
 "addresses": [
   {
     "type": "home",
     "value": "123 Main St, City, Country"
   },
   {
     "type": "work",
     "value": "456 Office Rd, City, Country"
   }

>  ],

Reviewer: Did you test the address case?

      - **Simple Attributes** contain a single attribute.

        ```json
        {
            "devices": ["d1", "d2"]
        }
        ```

      - **Complex Attributes** contain multiple sub-attributes.

        ```json
        {
          "emails": [
            {
              "value": "kim@gmail.com",
              "primary": true
            },
            {
              "type": "work",
              "value": "kim@wso2.com"
            }
          ]
        }
        ```

!!! info
    The following references provide comprehensive information about SCIM attribute types and their respective definitions. These details can help identify the type of attributes used in SCIM 2.0 payloads:

      - For attributes under core schema, user schema and SCIM2 specification-defined enterprise schema, refer to [RFC 7643 Section 8.7.1](https://datatracker.ietf.org/doc/html/rfc7643#section-8.7.1).

      {% if product_name == "WSO2 Identity Server" %}
      - For {{product_name}}-defined enterprise schema attributes, refer to the `scim2-schema-extension.config` file located in the `<IS_HOME>/repository/conf/` directory.
      {% endif %} 

>       - For custom schema attributes, check the `type` meta attribute of the mapped local attribute.

Reviewer: ```suggestion
      - For custom schema attributes, check the `dataType` meta attribute of the mapped local attribute.
```

Reviewer: https://is.docs.wso2.com/en/6.1.0/references/extend/provisioning/add-scim2-custom-user-schema-support/#add-custom-schema-dialect:~:text=Additional%20Properties%20element.-,dataType,-multiValued
